### DISCUSSION TITLE
VRP: is there a way to force a vehicle to visit certain nodes after visting specific nodes?

### DISCUSSION THREAD

**Discussion by fuegosour24** [google/or-tools#2684] [Routing (and legacy CP) questions]
Hi, 

I am working on a VRP scenario that has one depot(end location) and various factories (pick up locations) and drop off locations. A vehicle has to go from the factory to the drop off and back to the factory. As such, I have made various dummy variables to be able to perform this. For example, I have duplicated the "Factory 1" and "Factory 2" nodes in order to force them to be the starting locations (i.e. prevent drop off nodes from being start nodes as you can't pick up a product from the drop off location). After, in order to take advantage of the "pickup deliveries" feature, I have duplicated "factory/drop off" node sets so that a vehicle goes from the factory to the drop off and back to the factory. (i.e. [2,3],[5,4] where 2 = 4 and 3=5). 

However, this only works if there are a unique set of factory/drop off combinations. 

For instance, when I add a new route using an already existing pickup/factory location, the output solution goest from:
11(factory 1, starting location) --> 3(factory 1) --> 14(drop off 3) -->12 (drop off 3) --> 6 (drop off 2) --> 4(drop off). 

I was wondering if there was a way to add a conditional constraint (using penalties or constraints) where if a vehicle is at a certain node, it can't visit a list of other nodes? For instance, if a vehicle is at a drop off node, it can't go to another drop off node, it can only go to a pickup/factory node. 

```python
    data['distance_matrix'] = [
        
        #STARTING LOCATIONS
        [   0,  430,  299,  430,  377,  430,  377,  299,  471,  299,  471,430, 1467,  430, 1467],#DEPOT 
       [ 430,    0,  384,    0,  612,    0,  612,  384,  887,  384,  887, 0, 1744,    0, 1744], #FACTORY 1
       [ 299,  384,    0,  384,  666,  384,  666,    0,  727,    0,  727,384, 1765,  384, 1765],#FACTORY 2
        
        
        #FACTORY TO DROP OFF
       [ 430,    0,  384,    0,  612,    0,  612,  384,  887,  384,  887, 0, 1744,    0, 1744],#FACTORY 1
       [ 377,  612,  666,  612,    0,  612,    0,  666,  437,  666,  437,  612, 1138,  612, 1138],#DROP OFF 1
       #DROP OFF BACK TO FACTORY
        [ 430,    0,  384,    0,  612,    0,  612,  384,  887,  384,  887, 0, 1744,    0, 1744],#FACTORY 1
       [ 377,  612,  666,  612,    0,  612,    0,  666,  437,  666,  437,612, 1138,  612, 1138],#DROP OFF 1
       
        
        
        #FACTORY 2 TO DROP OFF #2
        [ 299,  384,    0,  384,  666,  384,  666,    0,  727,    0,  727,384, 1765,  384, 1765],#FACTORY 2
       [ 471,  887,  727,  887,  437,  887,  437,  727,    0,  727,    0,887, 1112,  887, 1112],#DROP OFF 2
        #DROP OFF #2 TO FACTORY 2
       [ 299,  384,    0,  384,  666,  384,  666,    0,  727,    0,  727,384, 1765,  384, 1765],#FACTORY 2
       [ 471,  887,  727,  887,  437,  887,  437,  727,    0,  727,    0,887, 1112,  887, 1112],#DROP OFF 2
        
        
        #FACTORY 1 TO DROP OFF#3
       [ 430,    0,  384,    0,  612,    0,  612,  384,  887,  384,  887,0, 1744,    0, 1744], #FACTORY 1
       [1467, 1744, 1765, 1744, 1138, 1744, 1138, 1765, 1112, 1765, 1112,1744,    0, 1744,    0],#DROP OFF 3
        
        #DROP OFF #3 TO FACTORY 1
       [ 430,    0,  384,    0,  612,    0,  612,  384,  887,  384,  887,0, 1744,    0, 1744],#FACTORY 1
       [1467, 1744, 1765, 1744, 1138, 1744, 1138, 1765, 1112, 1765, 1112,1744,    0, 1744,    0]] #DROP OFF 3



    data['pickups_deliveries'] = [
        [3,4],#pick up at F1, drop 1
        [6, 5], #drop 1 to F1
        [7,8],#pick up at F2 to drop 2
        [10,9],# drop 2 to F2
        [11,12],#pick up at F1 to drop 3
        [14,13]#Drop 3 to F1
    ]
    # [END pickups_deliveries]
    data['num_vehicles'] = 2
    data['starts'] = [1,2]
    data['ends'] = [0,0]
```
_____________________
OUTPUT
```sh
Objective: 398077
Route for vehicle 0:
 1 ->  11 ->  3 ->  14 ->  12 ->  6 ->  4 ->  13 ->  5 -> 0
Distance of the route: 3924 miles

Route for vehicle 1:
 2 ->  7 ->  10 ->  8 ->  9 -> 0
Distance of the route: 1753 miles

Total Distance of all routes: 5677 miles
```

**Comment by fuegosour24** [comment]
In summary, is there a way to constraint routes, not individual nodes ?

**Reply by Mizux** [reply]
you can't constrain "route" but for each node you can control which node next is allowed using  `routing.NextVar(index)`

see: https://groups.google.com/g/or-tools-discuss/c/Vz4sJtwBM2Q/m/XirUtGPzAwAJ
